File: frontend/lib/tinyfish-client.ts
/**
 * TinyFish Web Agent SSE client
 * Following cookbook patterns from https://github.com/tinyfish-io/tinyfish-cookbook
 *
 * Event types from TinyFish API:
 *   STARTED       - { runId, timestamp }
 *   STREAMING_URL - { streamingUrl } (live browser view, ~24h)
 *   PROGRESS      - { purpose } (action description)
 *   HEARTBEAT     - keep-alive
 *   COMPLETE      - { status: COMPLETED|FAILED|CANCELLED, resultJson, error }
 */

export interface TinyFishCallbacks {
  onStarted?: (runId: string) => void;
  onStreamUrl?: (url: string) => void;
  onStep?: (purpose: string) => void;
  onComplete?: (
    result: Record<string, unknown> | null,
    status: string
  ) => void;
  onError?: (error: string) => void;
}

export async function runTinyFishAutomation(
  url: string,
  goal: string,
  callbacks?: TinyFishCallbacks,
  browserProfile: "lite" | "stealth" = "lite"
): Promise<Record<string, unknown> | null> {
  const res = await fetch("/api/tinyfish", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ url, goal, browser_profile: browserProfile }),
  });

  if (!res.ok) {
    const err = await res.text();
    callbacks?.onError?.(err);
    return null;
  }

  const reader = res.body?.getReader();
  if (!reader) return null;

  const decoder = new TextDecoder();
  let result: Record<string, unknown> | null = null;
  // Buffer for handling chunked SSE lines (cookbook pattern)
  let buffer = "";

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split("\n");
    // Keep the last partial line in buffer
    buffer = lines.pop() ?? "";

    for (const line of lines) {
      if (!line.startsWith("data: ")) continue;

      try {
        const event = JSON.parse(line.slice(6));

        switch (event.type) {
          case "STARTED":
            callbacks?.onStarted?.(event.runId);
            break;
          case "STREAMING_URL":
            callbacks?.onStreamUrl?.(event.streamingUrl);
            break;
          case "PROGRESS":
            callbacks?.onStep?.(event.purpose || "Processing...");
            break;
          case "HEARTBEAT":
            // Keep-alive, no action needed
            break;
          case "COMPLETE":
            if (event.status === "COMPLETED") {
              result = event.resultJson || event.result || null;
              callbacks?.onComplete?.(result, event.status);
            } else {
              // FAILED or CANCELLED
              callbacks?.onError?.(
                event.error || `Automation ${event.status}`
              );
            }
            break;
        }
      } catch {
        // Skip malformed lines
      }
    }
  }

  return result;
}

/**
 * Run multiple TinyFish automations in parallel (cookbook pattern).
 * Returns settled results for each URL.
 */
export async function runTinyFishParallel(
  tasks: Array<{ url: string; goal: string }>,
  callbacks?: TinyFishCallbacks
): Promise<Array<PromiseSettledResult<Record<string, unknown> | null>>> {
  return Promise.allSettled(
    tasks.map((task) => runTinyFishAutomation(task.url, task.goal, callbacks))
  );
}

// Parse SSE event lines (shared helper)
export function parseSSELine(line: string): Record<string, unknown> | null {
  if (!line.startsWith("data: ")) return null;
  try {
    return JSON.parse(line.slice(6));
  } catch {
    return null;
  }
}

export function isCompleteEvent(event: Record<string, unknown>): boolean {
  return event.type === "COMPLETE" && event.status === "COMPLETED";
}

export function isErrorEvent(event: Record<string, unknown>): boolean {
  return (
    event.type === "COMPLETE" &&
    (event.status === "FAILED" || event.status === "CANCELLED")
  );
}

File: backend/app/services/tinyfish_client.py
"""TinyFish Web Agent API client - SSE streaming, sync, and async modes.

Following cookbook patterns from https://github.com/tinyfish-io/tinyfish-cookbook
API Docs: https://docs.mino.ai/

Endpoints:
  POST /run-sse   - SSE streaming (real-time UI)
  POST /run       - Synchronous (blocks until done)
  POST /run-async - Start, returns run_id for polling
  GET  /runs/{id} - Poll async status
"""
import httpx
import json
import logging
from typing import AsyncGenerator, Optional, Dict, Any

from app.config import settings

logger = logging.getLogger(__name__)

TINYFISH_BASE = settings.TINYFISH_BASE_URL
TINYFISH_HEADERS = {
    "X-API-Key": settings.TINYFISH_API_KEY,
    "Content-Type": "application/json",
}


def _build_payload(
    url: str,
    goal: str,
    stealth: bool = False,
    proxy_config: Optional[Dict] = None,
) -> Dict[str, Any]:
    """Build TinyFish API request payload."""
    payload: Dict[str, Any] = {
        "url": url,
        "goal": goal,
        "browser_profile": "stealth" if stealth else "lite",
    }
    if proxy_config:
        payload["proxy_config"] = proxy_config
    return payload


async def run_sse_automation(
    url: str,
    goal: str,
    stealth: bool = False,
    proxy_config: Optional[Dict] = None,
    timeout: int = 120,
) -> AsyncGenerator[Dict[str, Any], None]:
    """Stream TinyFish automation results via SSE.

    Yields event dicts with types: STARTED, STREAMING_URL, PROGRESS, HEARTBEAT, COMPLETE
    """
    payload = _build_payload(url, goal, stealth, proxy_config)

    try:
        async with httpx.AsyncClient(timeout=timeout) as client:
            async with client.stream(
                "POST",
                f"{TINYFISH_BASE}/run-sse",
                headers=TINYFISH_HEADERS,
                json=payload,
            ) as response:
                if response.status_code != 200:
                    error_body = ""
                    async for chunk in response.aiter_text():
                        error_body += chunk
                    logger.error(f"TinyFish SSE error {response.status_code}: {error_body}")
                    yield {"type": "COMPLETE", "status": "FAILED", "error": f"API error {response.status_code}"}
                    return

                buffer = ""
                async for chunk in response.aiter_text():
                    buffer += chunk
                    lines = buffer.split("\n")
                    buffer = lines[-1]  # keep incomplete line

                    for line in lines[:-1]:
                        if line.startswith("data: "):
                            try:
                                event = json.loads(line[6:])
                                yield event
                                if event.get("type") == "COMPLETE":
                                    return
                            except json.JSONDecodeError:
                                continue
    except httpx.TimeoutException:
        logger.error(f"TinyFish SSE timeout after {timeout}s for {url}")
        yield {"type": "COMPLETE", "status": "FAILED", "error": "Timeout"}
    except Exception as e:
        logger.error(f"TinyFish SSE error: {e}")
        yield {"type": "COMPLETE", "status": "FAILED", "error": str(e)}


def run_sync_automation(
    url: str,
    goal: str,
    stealth: bool = False,
    proxy_config: Optional[Dict] = None,
    timeout: int = 120,
) -> Dict[str, Any]:
    """Synchronous TinyFish call - blocks until completion. Use in Celery tasks or thread pool."""
    payload = _build_payload(url, goal, stealth, proxy_config)

    try:
        resp = httpx.post(
            f"{TINYFISH_BASE}/run",
            headers=TINYFISH_HEADERS,
            json=payload,
            timeout=timeout,
        )
        if resp.status_code != 200:
            logger.error(f"TinyFish sync error {resp.status_code}: {resp.text}")
            return {"status": "FAILED", "error": f"API error {resp.status_code}"}
        return resp.json()
    except httpx.TimeoutException:
        logger.error(f"TinyFish sync timeout for {url}")
        return {"status": "FAILED", "error": "Timeout"}
    except Exception as e:
        logger.error(f"TinyFish sync error: {e}")
        return {"status": "FAILED", "error": str(e)}


async def run_async_automation(
    url: str,
    goal: str,
    stealth: bool = False,
) -> Dict[str, Any]:
    """Start TinyFish automation asynchronously. Returns run_id for polling."""
    payload = _build_payload(url, goal, stealth)

    try:
        async with httpx.AsyncClient(timeout=30) as client:
            resp = await client.post(
                f"{TINYFISH_BASE}/run-async",
                headers=TINYFISH_HEADERS,
                json=payload,
            )
            if resp.status_code != 200:
                return {"error": f"API error {resp.status_code}"}
            return resp.json()
    except Exception as e:
        return {"error": str(e)}


async def poll_run_status(run_id: str) -> Dict[str, Any]:
    """Poll the status of an async TinyFish run."""
    try:
        async with httpx.AsyncClient(timeout=15) as client:
            resp = await client.get(
                f"https://agent.tinyfish.ai/v1/runs/{run_id}",
                headers=TINYFISH_HEADERS,
            )
            return resp.json()
    except Exception as e:
        return {"error": str(e)}
